### PR TITLE
feat(context-menu): add a simple context menu component

### PR DESCRIPTION
**Pull request by cal-smith**
Adds a simple context-menu implementation. [See here](https://react.carbondesignsystem.com/?path=/story/experimental-unstable-contextmenu--context-menu) for the existing react implementation this is based on. We deviate in a few ways, and add some component level styles to make things behave, but all the functionality should be the same.

This doesn't handle displaying on right click - We can add a service later to deal with that (I'm thinking something that could serve as a proper foundation for the dialog service ... Maybe we can address some of the positioning bugs at the same time).

### CODE REVIEW

File: src/context-menu/context-menu-item.component.ts [hunk 1]
```diff
@@ -0,0 +1,207 @@
+import {
+	Component,
+	HostBinding,
+	Input, Output,
```
Comment (zvonimirfras) [inline src/context-menu/context-menu-item.component.ts:4]:
```suggestion
	Input,
	Output,
```
?

File: src/context-menu/context-menu-item.component.ts [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+	// 		}
+	// 		nextElement.focus();
+	// 	}
+
+	// 	if (event.key === 'ArrowUp') {
+	// 		console.log('item up');
+	// 		(element.previousElementSibling as HTMLElement).focus();
+	// 	}
+	// }
+
```
Comment (zvonimirfras) [inline src/context-menu/context-menu-item.component.ts:198]:
```suggestion
```

or keeping for later?

File: src/context-menu/context-menu.component.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+		@SkipSelf() @Optional() protected parentContextMenu: ContextMenuComponent,
+		protected elementRef: ElementRef
+	) { }
+
+	ngOnChanges(changes: SimpleChanges) {
+		if (changes.open) {
+			if (changes.open.currentValue) {
+				this.focusMenu();
+			}
+		}
```
Comment (zvonimirfras) [inline src/context-menu/context-menu.component.ts:62]:
```suggestion
		if (changes.open && changes.open.currentValue) {
			this.focusMenu();
		}
```

File: src/context-menu/context-menu.component.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	@HostListener("keydown", ["$event"])
+	handleNavigation(event: KeyboardEvent) {
+		const subMenus = this.subMenus.toArray().filter(subMenu => subMenu !== this);
+		const menuItems = this.menuItems.toArray().filter(menuItem => {
+			// if the menu item is contained within any submenu then remove it from the array
+			return !subMenus.some(subMenu => {
+				// check if the menu item is contained within a submenu
+				return subMenu.menuItems.some(subMenuItem => {
+					return menuItem === subMenuItem;
+				});
```
Comment (zvonimirfras) [inline src/context-menu/context-menu.component.ts:93]:
```suggestion
				return subMenu.menuItems.some(subMenuItem => menuItem === subMenuItem);
```

File: src/context-menu/context-menu.component.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+			});
+		});
+		const currentIndex = menuItems.findIndex(menuItem => menuItem.tabindex === 0);
+		const list: HTMLElement = this.elementRef.nativeElement.querySelector("ul");
+
+		if (event.key === "ArrowDown") {
```
Comment (zvonimirfras) [inline src/context-menu/context-menu.component.ts:99]: Looks like you're trying to create a switch-case type flow with ifs on event.key. Maybe just use switch?

**Review by zvonimirfras — CHANGES_REQUESTED** [review]

File: src/context-menu/context-menu-item.component.ts (hunk 2 above)
Comment (cal-smith) [reply src/context-menu/context-menu-item.component.ts:199]: Nope! Missed that one!

File: src/context-menu/context-menu.component.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+					menuItems[menuItems.length - 1].focus();
+				} else {
+					if (currentIndex !== -1 && currentIndex > 0) {
+						menuItems[currentIndex - 1].focus();
+					}
+				}
+				break;
+			}
+			case "ArrowRight": {
+
```
Comment (zvonimirfras) [inline src/context-menu/context-menu.component.ts:90]:
```suggestion
```

**Review by zvonimirfras — DISMISSED** [review]
Just this line and we're good to go.

**Pull request merged**